tf_encrypted/keras/layers/dense.py

```diff
@@ -0,0 +1,118 @@
+      bias_regularizer: Regularizer function applied to the bias vector.
+      activity_regularizer: Regularizer function applied to
+          the output of the layer (its "activation")..
+      kernel_constraint: Constraint function applied to
+          the `kernel` weights matrix.
+      bias_constraint: Constraint function applied to the bias vector.
+
+  Input shape:
+      2D tensor with shape: `(batch_size, input_dim)`.
+
+  Output shape:
+      2D tensor with shape: `(batch_size, units)`.
+  """
+
+  def __init__(self,
+               units,
+               activation=None,
+               use_bias=True,
+               kernel_initializer='glorot_uniform',
+               bias_initializer='zeros',
+               kernel_regularizer=None,
+               bias_regularizer=None,
+               activity_regularizer=None,
+               kernel_constraint=None,
+               bias_constraint=None,
+               **kwargs):
+
+    super(Dense, self).__init__(**kwargs)
+
+    self.units = int(units)
+    self.activation = activations.get(activation)
+    self.use_bias = use_bias
+
+    self.kernel_initializer = initializers.get(kernel_initializer)
+    self.bias_initializer = initializers.get(bias_initializer)
+
+    if kernel_regularizer:
+      raise NotImplementedError(arg_not_impl_msg.format("kernel_regularizer",
+                                                        "Dense"))
+    if bias_regularizer:
+      raise NotImplementedError(arg_not_impl_msg.format("bias_regularizer",
+                                                        "Dense"))
+    if activity_regularizer:
+      raise NotImplementedError(arg_not_impl_msg.format("activity_regularizer",
+                                                        "Dense"))
+    if kernel_constraint:
+      raise NotImplementedError(arg_not_impl_msg.format("kernel_constraint",
+                                                        "Dense"))
+    if bias_constraint:
+      raise NotImplementedError(arg_not_impl_msg.format(" bias_constraint",
+                                                        "Dense"))
+
+  def compute_output_shape(self, input_shape):
+    return [input_shape[0], self.units]
+
+  def build(self, input_shape):
+    units_in = int(input_shape[1])
+    kernel = self.kernel_initializer([units_in,
+                                      self.units])
+    self.kernel = self.prot.define_private_variable(kernel)
+
+    if self.use_bias:
+      bias = self.bias_initializer([self.units])
+      self.bias = self.prot.define_private_variable(bias)
+    else:
+      self.bias = None
+
+    self.built = True
+
+  def call(self, inputs):
+
+    input_shape = inputs.shape.as_list()
+
+    #  if the input to the layer has a rank greater than 2,
+    #  then it is flattened prior to the initial dot product with kernel.
+    if len(input_shape) > 2:
+      flat_shape = [input_shape[0]] + [np.prod(input_shape[1:])]
+      inputs = inputs.reshape(flat_shape)
```

Comment:
you should be able to simply use `flat_shape = [input_shape[0], -1]` here, avoiding the numpy import

Reply:
ah yes! sorry about that.